OSIDB-3338: Create djangoql suggestions endpoints

MrMarble opened the pull request:
This pull request introduces new API endpoints for DjangoQL and includes several related changes across multiple files. 

### New API Endpoints:

* **FlawIntrospectionView**: This endpoint return a json of all the models (included in the schema) with the type of the fields and the possible values

* **FlawSuggestionsView**: This endpoint returns the possible values for a given field specified in the schema, for giving suggestions to the user while typing


>[!NOTE]
>I have chosen the models and fields a bit “blindly”, it may be necessary to discuss which fields need suggestions and which models we allow for the search.

Closes OSIDB-3338

osoukup commented on osidb/filters.py:
```diff
@@ -260,6 +260,30 @@ class FlawQLSchema(DjangoQLSchema):
+            "workflow_state",
+        ],
+        Affect: ["resolution", "affectedness", "impact"],
+        Tracker: ["type", "status"],
```
When `status` I would expect `resolution` too :thinking:

osoukup commented on osidb/filters.py:
```diff
@@ -260,6 +260,30 @@ class FlawQLSchema(DjangoQLSchema):
+        Affect: ["resolution", "affectedness", "impact"],
+        Tracker: ["type", "status"],
+        FlawReference: ["type"],
+        FlawCVSS: ["issuer"],
```
I would also add `version`

osoukup commented on osidb/filters.py:
```diff
@@ -260,6 +260,30 @@ class FlawQLSchema(DjangoQLSchema):
+            "impact",
+            "requires_cve_description",
+            "major_incident_state",
+            "workflow_state",
```
The order of the items seems quite random. If there is no rule to sort I would prefer the alphabet then

osoukup commented on osidb/filters.py:
```diff
@@ -260,6 +260,30 @@ class FlawQLSchema(DjangoQLSchema):
+        Flaw: [
+            "source",
+            "impact",
+            "requires_cve_description",
```
NIST validation flag seems to me quite similar so I would expect that one too

osoukup reviewed (approved):
I have some minor suggestions but generally LGTM

MrMarble commented:
> I have some minor suggestions but generally LGTM

Sorted all the keys and added the suggestions

As for the `Include` models, I added only the one that are directly  related to the Flaw

![image](https://github.com/user-attachments/assets/e265b3cf-613d-4ea0-8b5e-ede67549b307)
> Screenshot of the introspection endpoint filtered to shown only model relations

osoukup reviewed (approved):
LGTM

MrMarble commented on docs/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Added
 - Add CVEorg collector (OSIDB-2234)
 - Allow trackers to have manually set SLAs (OSIDB-3374)
+- Create new API endpoints for DjangoQL (OSIDB-3338)
```
Updated with master and also added `flaw.components`, `flaw.owner`, `affect.ps_module` and `affect.ps_component` to the lists of suggestions as requested by Jessie

jobselko replied:
Your change is still in `4.3.0` release section :)

MrMarble replied:
fixed

jobselko commented on osidb/filters.py:
```diff
@@ -260,6 +260,33 @@ class FlawQLSchema(DjangoQLSchema):
     any field in the model, which is not desirable in this case.
     """
 
+    include = (
+        Flaw,
+        Affect,
+        Tracker,
+        Package,
+        FlawCVSS,
+        FlawAcknowledgment,
+        FlawReference,
+    )
+
+    suggest_options = {
+        Flaw: [
+            "impact",
+            "major_incident_state",
+            "nist_cvss_validation",
+            "requires_cve_description",
+            "source",
+            "workflow_state",
+            "components",
+            "owner",
+        ],
+        Affect: ["affectedness", "impact", "resolution", "ps_module", "ps_component"],
+        Tracker: ["resolution", "status", "type"],
+        FlawReference: ["type"],
+        FlawCVSS: ["issuer", "version"],
+    }
```
Would it be possible to sort everything alphabetically?

MrMarble replied:
sorted

jobselko replied:
Please look on items in `suggest_options` (i.e. `Affect`, `Flaw`, ...) and then on items in `Affect`.

MrMarble replied:
Third time's a charm
